surefire-providers/common-junit48/src/main/java/org/apache/maven/surefire/common/junit48/GroupMatcherCategoryFilter.java

```diff
@@ -63,33 +63,50 @@
     @Override
     public boolean shouldRun( Description description )
     {
-        if ( description.getMethodName() == null || description.getTestClass() == null )
-        {
+        if(invalidTestClass(description)){
             return shouldRun( description, null, null );
         }
-        else
+
+        if (describesTestClass(description)) // is a test class
+        {
+            Class<?> testClass = description.getTestClass();
+            return shouldRun( description, null, testClass );
+        }
+        else // is a test method
         {
             Class<?> testClass = description.getTestClass();
             return shouldRun( description, createSuiteDescription( testClass ), testClass );
         }
     }
 
+    private boolean describesTestClass(Description description) {
+        // Description parser in Junit 4.8 can return "null" String.
+        return description.getMethodName() == null || description.getMethodName().equals("null");
+    }
+
+    private boolean invalidTestClass(Description description) {
+        return description.getTestClass() == null;
+    }
+
     private static void findSuperclassCategories( Set<Class<?>> cats, Class<?> clazz )
     {
-        if ( clazz != null && clazz.getSuperclass() != null )
+        if (hasSuperclass(clazz))
         {
             Category cat = clazz.getSuperclass().getAnnotation( Category.class );
             if ( cat != null )
             {
+                // Found categories in current superclass
                 addAll( cats, cat.value() );
             }
-            else
-            {
-                findSuperclassCategories( cats, clazz.getSuperclass() );
-            }
+            // Search the hierarchy
+            findSuperclassCategories( cats, clazz.getSuperclass() );
```

Comment:
This canot be without the original `else` because the Category always overrides category. There is no merge in the JUnit!

Reply:
Hi, thanks for the Review! However, I don't understand this comment. Do you mean that a @Category in a subclass should override (i.e. replace) the classes in the @Category from the superclass? Because this is exactly the opposite of what I implemented.  I expect that the classes in the superclass are included as categories of the subclass, without needing to re-declare them.

Reply:
https://github.com/apache/maven-surefire/pull/248/#discussion_r328453423
Yes, exactly!
If you have any questions, feel free to ask. We will help.

Reply:
@YiannisDermitzakis
We spoke with @IoannisDermitzakis about annotation `@Inherited`. Can you join us and adapt the code? Thx

Reply:
Hi @Tibor17, I am halfway through the changes, but my time is currently very limited. I estimate I'll post an updated PR within the next two weeks.

Reply:
@YiannisDermitzakis
@IoannisDermitzakis
We will cut a new release version. If you wan to have this feature released soon, pls come back to the discussion. Thx